### ISSUE TITLE
Spack v0.16.2 can't download using Python 3.8.11 

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#24716]
### Steps to reproduce the issue

```console
$ spack install zlib
==> [2021-07-05-19:56:17.685907] Error: Failed to install zlib due to AttributeError: 'NoneType' object has no attribute 'replace'
Traceback (most recent call last):
  File "./spack/bin/spack", line 66, in <module>
    sys.exit(spack.main.main())
  File "/root/spack/lib/spack/spack/main.py", line 762, in main
    return _invoke_command(command, parser, args, unknown)
  File "/root/spack/lib/spack/spack/main.py", line 490, in _invoke_command
    return_val = command(parser, args)
  File "/root/spack/lib/spack/spack/cmd/install.py", line 365, in install
    install_specs(args, kwargs, zip(abstract_specs, specs))
  File "/root/spack/lib/spack/spack/cmd/install.py", line 207, in install_specs
    builder.install()
  File "/root/spack/lib/spack/spack/installer.py", line 1535, in install
    self._install_task(task)
  File "/root/spack/lib/spack/spack/installer.py", line 1101, in _install_task
    _install_from_cache(pkg, cache_only, explicit, unsigned,
  File "/root/spack/lib/spack/spack/installer.py", line 273, in _install_from_cache
    installed_from_cache = _try_install_from_binary_cache(
  File "/root/spack/lib/spack/spack/installer.py", line 394, in _try_install_from_binary_cache
    matches = binary_distribution.get_mirrors_for_spec(
  File "/root/spack/lib/spack/spack/binary_distribution.py", line 1396, in get_mirrors_for_spec
    results = try_direct_fetch(spec,
  File "/root/spack/lib/spack/spack/binary_distribution.py", line 1336, in try_direct_fetch
    buildcache_fetch_url = url_util.join(
  File "/root/spack/lib/spack/spack/util/url.py", line 156, in join
    obj = urllib_parse.urlparse(
  File "/home/runner/.setup-spack/._view/33gon3ice5endxsdau2xe2yv35p7kz2v/lib/python3.8/urllib/parse.py", line 376, in urlparse
    splitresult = urlsplit(url, scheme, allow_fragments)
  File "/home/runner/.setup-spack/._view/33gon3ice5endxsdau2xe2yv35p7kz2v/lib/python3.8/urllib/parse.py", line 433, in urlsplit
    scheme = _remove_unsafe_bytes_from_url(scheme)
  File "/home/runner/.setup-spack/._view/33gon3ice5endxsdau2xe2yv35p7kz2v/lib/python3.8/urllib/parse.py", line 422, in _remove_unsafe_bytes_from_url
    url = url.replace(b, "")
AttributeError: 'NoneType' object has no attribute 'replace'
```

## Information on your system

* **Spack:** 0.16.2
* **Python:** 3.8.11
* **Platform:** linux-ubuntu18.04-zen2

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
Will be fixed once #24686 is backported and we cut a 0.16.3 release.

**Comment by haampie** [comment]
Closing this since it'll be backported
